File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/RecipeService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.item.ItemDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.item.pantryitem.PantryItemDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.recipe.RecipeDetailDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.recipe.RecipeDetailWithUserInfoDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.recipe.RecipeListDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.recipe.RecipeCreateDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.recipe.RecipeGlobalListDto;
import at.ac.tuwien.sepr.groupphase.backend.entity.ApplicationUser;
import at.ac.tuwien.sepr.groupphase.backend.entity.Item;
import at.ac.tuwien.sepr.groupphase.backend.entity.Recipe;
import at.ac.tuwien.sepr.groupphase.backend.exception.AlreadyRatedException;

import java.util.List;

public interface RecipeService {


    /**
     * Find all items in a recipe where {@code description} is a substring of the item description ordered by the item id.
     *
     * @param description the description
     * @param recipeId    the recipe id
     * @return ordered list of all items in the recipe where {@code description} is a substring of the item description
     */
    List<Item> findItemsByDescription(String description, long recipeId);

    /**
     * Saves an item that belongs into a recipe.
     *
     * @param item     the item to save
     * @param recipeId the id of the corresponding recipe
     * @return the saved item
     */
    Item addItemToRecipe(Item item, long recipeId);

    /**
     * Deletes an item in a recipe.
     *
     * @param itemId   the id of the item to delete
     * @param recipeId the recipe id
     */
    void deleteItem(long recipeId, long itemId);


    /**
     * Create a new recipe.
     *
     * @param recipeDto the recipe to create
     * @return the created recipe with id
     */
    RecipeDetailDto createRecipe(RecipeCreateDto recipeDto);

    /**
     * Find a recipe by its id.
     *
     * @param id the id of the recipe
     * @return the recipe with the corresponding id
     */
    RecipeDetailDto getById(long id);

    /**
     * Find a recipe by its id and see if the user liked it or not.
     *
     * @param id   the id of the recipe
     * @param user the user to get the information from
     * @return the recipe with the corresponding id
     */
    RecipeDetailWithUserInfoDto getByIdWithInfo(long id, ApplicationUser user);


    /**
     * Return all recipes from the associated user.
     * Still needs doing
     *
     * @return the recipes from the specified user
     */
    List<RecipeListDto> getRecipesFromUser();

    /**
     * Update a recipe with the corresponding id.
     *
     * @param toUpdate the detailDto to update
     * @return the updated recipe
     */
    RecipeDetailDto updateRecipe(RecipeDetailDto toUpdate);

    /**
     * Get a List of all public RecipeListDto ordered in descending order by their like count.
     *
     * @return the list of all public recipes
     */
    List<RecipeGlobalListDto> getPublicRecipeOrderedByLikes(ApplicationUser user);

    /**
     * Delete a recipe from the database.
     *
     * @param id the id of the recipe to delete
     */
    void deleteRecipe(long id);

    /**
     * Updates an item in a recipe.
     *
     * @param item     the updated item
     * @param recipeId the id of the recipe, where the item is in
     * @return the updated item
     */
    Item updateItem(ItemDto item, long recipeId);

    /**
     * Like a recipe.
     * If a recipe is already disliked, remove the dislike
     *
     * @param recipeId Update the recipe to increase the like count
     * @param user     the user who sent the like
     */
    RecipeDetailDto likeRecipe(long recipeId, ApplicationUser user) throws AlreadyRatedException;

    /**
     * Dislike a recipe.
     * If a recipe is already liked, remove the like
     *
     * @param recipeId Update the recipe to increase the dislike count
     * @param user     the user who sent the dislike
     */
    RecipeDetailDto dislikeRecipe(long recipeId, ApplicationUser user) throws AlreadyRatedException;

    /**
     * Return a list of recipes from user whose name matches the searchparams.
     *
     * @param owner        the owner whose recipes should be returned
     * @param searchParams the string that should find a name
     * @return a list of all matching recipes
     */
    List<RecipeListDto> searchOwnRecipe(ApplicationUser owner, String searchParams);

    /**
     * Return a list of recipes the user liked whose name matches the searchparams.
     *
     * @param owner        the owner whose recipes should be returned
     * @param searchParams the string that should find a name
     * @return a list of all matching recipes
     */
    List<RecipeListDto> searchLikedRecipe(ApplicationUser owner, String searchParams);


    /**
     * Return a list of recipes whose name matches the searchparams.
     *
     * @param searchParams the string that should find a name
     * @return a list of all matching recipes
     */
    List<RecipeGlobalListDto> searchGlobalRecipe(ApplicationUser user, String searchParams);

    /**
     * Getting the recipes the user has liked.
     *
     * @param user the user
     * @return List of recipes from the user
     */
    List<Recipe> getLikedRecipesByUserEmail(ApplicationUser user);

}

File: frontend/src/app/services/recipe.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Globals} from '../global/globals';
import {
  RecipeCreateWithoutUserDto,
  RecipeDetailDto, RecipeDetailWithUserInfoDto,
  RecipeGlobalListDto,
  RecipeListDto,
  RecipeSearch
} from "../dtos/recipe";
import {ItemCreateDto, ItemDetailDto} from "../dtos/item";
import {AddItemToShoppingListDto, RemoveRecipeIngredientsFromPantryDto} from "../dtos/RecipePantryShoppingList";

@Injectable({
  providedIn: 'root'
})
export class RecipeService {

  private recipeBaseUri: string = this.globals.backendUri + '/group';

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  /**
   * Create a new recipe
   * @return the returned recipe
   * @param recipe
   */

  createRecipe(recipe: RecipeCreateWithoutUserDto): Observable<RecipeDetailDto> {
    return this.httpClient.post<RecipeDetailDto>(`${this.recipeBaseUri}/recipe/create`, recipe);
  }

  /**
   * Loads a recipe by its id.
   *
   * @param id id of the recipe to load
   */
  getRecipeById(id: number): Observable<RecipeDetailDto> {
    return this.httpClient.get<RecipeDetailDto>(`${this.recipeBaseUri}/${id}/recipe`);
  }


  /**
   * Loads a recipe by its id.
   *
   * @param id id of the recipe to load
   */
  getRecipeWithInfoById(id: number): Observable<RecipeDetailWithUserInfoDto> {
    return this.httpClient.get<RecipeDetailWithUserInfoDto>(`${this.recipeBaseUri}/${id}/recipe/info`);
  }

  /**
   * Filters for items in a recipe.
   *
   * @param id the recipe id
   * @param recipeSearch the search string
   */
  filterRecipe(id: number, recipeSearch: RecipeSearch): Observable<RecipeDetailDto> {
    let params = new HttpParams();
    params = params.append('details', recipeSearch.details)
    return this.httpClient.get<RecipeDetailDto>(`${this.recipeBaseUri}/${id}/recipe/search`, {params});
  }

  /**
   * Persists an item belonging to a recipe to the backend.
   *
   * @param id the recipe id
   * @param item to persist
   */
  createItem(id: number, item: ItemCreateDto): Observable<ItemDetailDto> {
    return this.httpClient.post<ItemDetailDto>(`${this.recipeBaseUri}/${id}/recipe`, item);
  }

  /**
   * Deletes an ingredient from a recipe
   *
   * @param recipeId the recipe id
   * @param id the item id
   */
  deleteIngredient(recipeId: number, id: number): Observable<ItemDetailDto> {
    return this.httpClient.delete<ItemDetailDto>(`${this.recipeBaseUri}/${recipeId}/recipe/${id}`);
  }


  /**
   * Get all recipes associated with the user that sends this request.
   * @return all recipes associated with the user
   */
  getRecipesFromUser(): Observable<RecipeListDto[]> {
    return this.httpClient.get<RecipeListDto[]>(`${this.recipeBaseUri}/recipe/list`);
  }

  /**
   * Get all recipes liked by the user that sends this request.
   * @return all recipes liked by the user
   */
  getLikedRecipesFromUser(): Observable<RecipeListDto[]> {
    return this.httpClient.get<RecipeListDto[]>(`${this.recipeBaseUri}/recipe/likedlist`);
  }

  /**
   * Update recipe with recipeId with new recipe.
   * @param toUpdate the RecipeDetailDto with new information
   * @return the updated recipeDetailDto
   */
  updateRecipe(toUpdate: RecipeDetailDto): Observable<RecipeDetailDto> {
    return this.httpClient.put<RecipeDetailDto>(`${this.recipeBaseUri}/recipe/update`, toUpdate);
  }

  /**
   * Get the list of all public recipes ordered by their like count (desc.).
   * @return the list of all public recipes
   */
  getPublicRecipeOrderedByLikes(): Observable<RecipeGlobalListDto[]> {
    return this.httpClient.get<RecipeGlobalListDto[]>(`${this.recipeBaseUri}/recipe/global`);
  }

  /**
   * Delete a recipe with the given id.
   * @param id the id of the recipe to delete
   */
  deleteRecipe(id: number): Observable<void> {

    return this.httpClient.delete<void>(`${this.recipeBaseUri}/recipe/${id}/delete`);

  }

  /**
   * Increase the like count of the recipe by 1 and store who liked it.
   * If the user already disliked, remove the dislike
   * @param id of the recipe to like
   */
  likeRecipe(id: number): Observable<RecipeDetailDto> {
    return this.httpClient.put<RecipeDetailDto>(`${this.recipeBaseUri}/recipe/${id}/like`, {});
  }

  /**
   * Increase the dislike count of the recipe by 1 and store who disliked it.
   * If the user already liked, remove the like
   * @param id of the recipe to dislike
   */
  dislikeRecipe(id: number): Observable<RecipeDetailDto> {
    return this.httpClient.put<RecipeDetailDto>(`${this.recipeBaseUri}/recipe/${id}/dislike`, {});
  }

  /**
   * Search for all public recipes with a search string
   * @param searchParam the string to search for
   */
  searchGlobalRecipes(searchParam: RecipeSearch): Observable<RecipeGlobalListDto[]> {
    let params = new HttpParams();
    params = params.append('details', searchParam.details);
    return this.httpClient.get<RecipeGlobalListDto[]>(`${this.recipeBaseUri}/recipe/search/global`, {params});
  }

  /**
   * Search for all public recipes with a search string
   * @param searchParam the string to search for
   */
  searchLikedRecipes(searchParam: RecipeSearch): Observable<RecipeListDto[]> {
    let params = new HttpParams();
    params = params.append('details', searchParam.details);
    return this.httpClient.get<RecipeListDto[]>(`${this.recipeBaseUri}/recipe/search/liked`, {params});
  }

  /**
   * Search for all owned recipes with a search string
   * @param searchParam the string to search for
   */
  searchOwnRecipes(searchParam: RecipeSearch): Observable<RecipeListDto[]> {
    let params = new HttpParams();
    params = params.append('details', searchParam.details)
    return this.httpClient.get<RecipeListDto[]>(`${this.recipeBaseUri}/recipe/search/own`, {params});
  }

  /**
   * Update an item in a recipe.
   * @param item the item to update
   * @param recipeId the id of the recipe the item belongs to
   */
  updateItemInRecipe(item: ItemDetailDto, recipeId:number): Observable<ItemDetailDto> {
    return this.httpClient.put<ItemDetailDto>(`${this.recipeBaseUri}/${recipeId}/recipe`, item);
  }

  removeRecipeIngredientsFromPantry(recipeId:number, pantryId:number , portion:number): Observable<RemoveRecipeIngredientsFromPantryDto> {

    return this.httpClient.put<RemoveRecipeIngredientsFromPantryDto>(`${this.recipeBaseUri}/recipe/${recipeId}/pantry/${pantryId}/${portion}`, {});
  }


  /**
   * Return a suggestion for the user what items he can add to his shopping list.
   * This function takes into account what already is in the selected shopping list and the optional pantry
   *
   * @param recipeId       the recipe with the ingredients
   * @param shoppingListId the shopping list to add
   * @param pantryId       the pantry the user wants to be considered
   * @return a list of items with
   */
  selectIngredientsForShoppingListWithPantry(recipeId: number, shoppingListId: number, pantryId: number): Observable<AddItemToShoppingListDto> {
    return this.httpClient.get<AddItemToShoppingListDto>(`${this.recipeBaseUri}/recipe/${recipeId}/shoppinglist/${shoppingListId}/pantry/${pantryId}`);
  }

  selectIngredientsForShoppingList(recipeId: number, shoppingListId: number): Observable<AddItemToShoppingListDto> {
    return this.httpClient.get<AddItemToShoppingListDto>(`${this.recipeBaseUri}/recipe/${recipeId}/shoppinglist/${shoppingListId}`);
  }

  exportRecipe(recipeId:number) {
    return this.httpClient.get(`${this.globals.backendUri}/recipe/${recipeId}/pdf`, {responseType: "blob"});
  }
}
